experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp

```diff
@@ -71,6 +83,187 @@ LogicalResult HandshakeSpeculationPass::placeUnits(Value ctrlSignal) {
   return success();
 }
 
+// This recursive function traverses the IR and creates a control path
+// by replicating the branches along the way. It stops at commits and
+// connects them to the newly created control path, with value ctrlSignal
+bool HandshakeSpeculationPass::routeBranchControlTraversal(
+    llvm::DenseSet<Operation *> &visited, Value ctrlSignal, Operation *currOp) {
+  // End traversal if already visited
+  if (visited.contains(currOp))
+    return false;
+  visited.insert(currOp);
+
+  bool found_commit = false;
+  if (isa<handshake::SpecCommitOp>(currOp)) {
+    // Connect commit to the correct control signal and end traversal
+    currOp->setOperand(1, ctrlSignal);
+    found_commit = true;
+  } else if (isa<handshake::ConditionalBranchOp>(currOp)) {
+    // Replicate a branch in the control path and use new control signal
+    MLIRContext *ctx = &getContext();
+    OpBuilder builder(ctx);
+
+    builder.setInsertionPointAfterValue(ctrlSignal);
+
+    // The Speculating Branch will discard the control token if the token
+    // in the speculative path is non-speculative.
+    Operation *branchDisc = builder.create<handshake::SpeculatingBranchOp>(
+        currOp->getLoc(), currOp->getResult(0), // specTagOperand
```

Comment:
I though this was wrong on my first read through, but is actually a pretty elegant way to solve the complexity of the routing- well done.

Reply:
I would include why this works in the presentation (guaranteed synchronicity of the condition and control tokens) since I didn't get this at all from Haoran's thesis.

Reply:
If feasible and sensible, a note in the code would also help future developers :)